Docker daemon crashes in debug mode (-D)

<!--
If you are reporting a new issue, make sure that we do not have any duplicates
already open. You can ensure this by searching the issue list for this
repository. If there is a duplicate, please close your issue and add a comment
to the existing issue instead.

If you suspect your issue is a bug, please edit your issue description to
include the BUG REPORT INFORMATION shown below. If you fail to provide this
information within 7 days, we cannot debug your issue and will close it. We
will, however, reopen it if you later provide the information.

For more information about reporting issues, see
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#reporting-other-issues

---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------
Use the commands below to provide key information from your environment:
You do NOT have to include this information if this is a FEATURE REQUEST
-->

**Output of `docker version`:**

```
Client:
 Version:      1.12.0-rc4
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   e4a0dbc
 Built:        Wed Jul 13 03:39:43 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-rc4
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   e4a0dbc
 Built:        Wed Jul 13 03:39:43 2016
 OS/Arch:      linux/amd64
```


**Output of `docker info`:**

```
Containers: 6
 Running: 0
 Paused: 0
 Stopped: 6
Images: 131
Server Version: 1.12.0-rc4
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 213
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local vmdk
 Network: overlay host null bridge
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 3.13.0-85-generic
Operating System: Ubuntu 14.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.955 GiB
Name: hte-1s-eng-dhcp98
ID: H3SC:SAIV:HFHW:WYI5:WC4O:GEVQ:S6AJ:VIJ2:4CVT:UXDQ:T4GW:7MCA
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 15
 Goroutines: 25
 System Time: 2016-07-25T23:46:32.0427493-07:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Running docker 1.12 RC4 on Ubuntu.
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="14.04.1 LTS, Trusty Tahr"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 14.04.1 LTS"
VERSION_ID="14.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"

**Steps to reproduce the issue:**
1. Start docker in debug mode (-D) - dockerd -D --dns 10.112.0.1 --dns 10.112.64.1 --dns 10.112.64.2 --debug=true --log-level=debug --raw-logs
2. Assume there is already a volume created try inspect on the volume or run a container with that volume
3. Docker daemon crashes (messages below)


**Describe the results you received:**
time="2016-07-25T23:13:33.284121156-07:00" level=debug msg="Registering POST, /networks/{id:.*}/connect"
time="2016-07-25T23:13:33.284385589-07:00" level=debug msg="Registering POST, /networks/{id:.*}/disconnect"
time="2016-07-25T23:13:33.284646449-07:00" level=debug msg="Registering DELETE, /networks/{id:.*}"
time="2016-07-25T23:13:33.286357011-07:00" level=info msg="API listen on /var/run/docker.sock"
time="2016-07-25T23:13:42.580250743-07:00" level=debug msg="Calling POST /v1.24/containers/create"
time="2016-07-25T23:13:42.581383267-07:00" level=debug msg="form data: {\"AttachStderr\":true,\"AttachStdin\":true,\"AttachStdout\":true,\"Cmd\":null,\"Domainname\":\"\",\"Entrypoint\":null,\"Env\":[],\"HostConfig\":{\"AutoRemove\":false,\"Binds\":[\"dvol_3:/mnt/myvol\"],\"BlkioDeviceReadBps\":null,\"BlkioDeviceReadIOps\":null,\"BlkioDeviceWriteBps\":null,\"BlkioDeviceWriteIOps\":null,\"BlkioWeight\":0,\"BlkioWeightDevice\":null,\"CapAdd\":null,\"CapDrop\":null,\"Cgroup\":\"\",\"CgroupParent\":\"\",\"ConsoleSize\":[0,0],\"ContainerIDFile\":\"\",\"CpuCount\":0,\"CpuPercent\":0,\"CpuPeriod\":0,\"CpuQuota\":0,\"CpuShares\":0,\"CpusetCpus\":\"\",\"CpusetMems\":\"\",\"Devices\":[],\"DiskQuota\":0,\"Dns\":[],\"DnsOptions\":[],\"DnsSearch\":[],\"ExtraHosts\":null,\"GroupAdd\":null,\"IOMaximumBandwidth\":0,\"IOMaximumIOps\":0,\"IpcMode\":\"\",\"Isolation\":\"\",\"KernelMemory\":0,\"Links\":null,\"LogConfig\":{\"Config\":{},\"Type\":\"\"},\"Memory\":0,\"MemoryReservation\":0,\"MemorySwap\":0,\"MemorySwappiness\":-1,\"NetworkMode\":\"default\",\"OomKillDisable\":false,\"OomScoreAdj\":0,\"PidMode\":\"\",\"PidsLimit\":0,\"PortBindings\":{},\"Privileged\":false,\"PublishAllPorts\":false,\"ReadonlyRootfs\":false,\"RestartPolicy\":{\"MaximumRetryCount\":0,\"Name\":\"no\"},\"SecurityOpt\":null,\"ShmSize\":0,\"UTSMode\":\"\",\"Ulimits\":null,\"UsernsMode\":\"\",\"VolumeDriver\":\"\",\"VolumesFrom\":null},\"Hostname\":\"\",\"Image\":\"busybox\",\"Labels\":{},\"NetworkingConfig\":{\"EndpointsConfig\":{}},\"OnBuild\":null,\"OpenStdin\":true,\"StdinOnce\":true,\"Tty\":true,\"User\":\"\",\"Volumes\":{},\"WorkingDir\":\"\"}"
time="2016-07-25T23:13:42.629342113-07:00" level=debug msg="Getting volume reference for name: dvol_3"
time="2016-07-25T23:13:42.629525964-07:00" level=debug msg="Probing all drivers for volume with name: dvol_3"
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x8 pc=0xc2c6e1]

goroutine 134 [running]:
panic(0x1af0ac0, 0xc82000e0f0)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/docker/docker/pkg/plugins.readPluginJSONInfo(0xc82036aaf4, 0x5, 0xc82036ae80, 0x1e, 0x0, 0x0, 0x0)
        /go/src/github.com/docker/docker/pkg/plugins/discovery.go:119 +0x221
github.com/docker/docker/pkg/plugins.(*localRegistry).Plugin(0xc820047e70, 0xc82036aaf4, 0x5, 0xc82004a1c0, 0x0, 0x0)
        /go/src/github.com/docker/docker/pkg/plugins/discovery.go:80 +0xa65
github.com/docker/docker/pkg/plugins.loadWithRetry(0xc82036aaf4, 0x5, 0xc820a82f00, 0xc8203b8240, 0x0, 0x0)
        /go/src/github.com/docker/docker/pkg/plugins/plugins.go:167 +0xa2
github.com/docker/docker/pkg/plugins.GetAll.func1(0xc820a82fb0, 0xc820a8be00, 0xc82036aaf4, 0x5)
        /go/src/github.com/docker/docker/pkg/plugins/plugins.go:250 +0x5f
created by github.com/docker/docker/pkg/plugins.GetAll
        /go/src/github.com/docker/docker/pkg/plugins/plugins.go:252 +0x5a5


**Describe the results you expected:**

Expected to get docker to provide volume data

**Additional information you deem important (e.g. issue happens only occasionally):**

--- a/pkg/plugins/discovery.go
+++ b/pkg/plugins/discovery.go
@@ -116,7 +116,7 @@ func readPluginJSONInfo(name, path string) (*Plugin, error) {
 		return nil, err
 	}
 	p.name = name
-	if len(p.TLSConfig.CAFile) == 0 {
+	if p.TLSConfig != nil && len(p.TLSConfig.CAFile) == 0 {
 		p.TLSConfig.InsecureSkipVerify = true
 	}
 	p.activateWait = sync.NewCond(&sync.Mutex{})

--- a/pkg/plugins/discovery_test.go
+++ b/pkg/plugins/discovery_test.go
@@ -117,3 +117,36 @@ func TestFileJSONSpecPlugin(t *testing.T) {
 		t.Fatalf("Expected plugin Key `/usr/shared/docker/certs/example-key.pem`, got %s\n", plugin.TLSConfig.KeyFile)
 	}
 }
+
+func TestFileJSONSpecPluginWithoutTLSConfig(t *testing.T) {
+	tmpdir, unregister := Setup(t)
+	defer unregister()
+
+	p := filepath.Join(tmpdir, "example.json")
+	spec := `{
+  "Name": "plugin-example",
+  "Addr": "https://example.com/docker/plugin"
+}`
+
+	if err := ioutil.WriteFile(p, []byte(spec), 0644); err != nil {
+		t.Fatal(err)
+	}
+
+	r := newLocalRegistry()
+	plugin, err := r.Plugin("example")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if plugin.name != "example" {
+		t.Fatalf("Expected plugin `plugin-example`, got %s\n", plugin.Name)
+	}
+
+	if plugin.Addr != "https://example.com/docker/plugin" {
+		t.Fatalf("Expected plugin addr `https://example.com/docker/plugin`, got %s\n", plugin.Addr)
+	}
+
+	if plugin.TLSConfig != nil {
+		t.Fatalf("Expected plugin TLSConfig nil, got %v\n", plugin.TLSConfig)
+	}
+}

